perlfunc/pack: fix pack @ example, indentation

mauke opened the pull request:
- Consistently use 4-space indentation for all verbatim blocks (not 1 space, 3 spaces, 7 spaces, or tabs).
- Fix the alignment of the "This code:" ... "gives this result:" columns. (I had to change the sample strings slightly to fit in the line length limit imposed by t/porting/podcheck.t.)
- Fix the `@1A((@2A)@3A)` example to use single quotes, not double quotes (which would interpolate array variables).
- Replace `pack L=>0x12345678` by the more obvious `pack("L", 0x12345678)` (we're not trying to obfuscate our example code (right?)).

Fixes #22425.

mauke commented:
The 3b10bc6 commit only touched the documentation in perlfunc. No actual code was broken.

More specifically, Tom Christiansen changed the code examples in the documentation to consistently use double quotes for strings. This was mostly fine, except for this one example in `perldoc -f pack`:

> Therefore, the result of
> ```perl
> pack("@1A((@2A)@3A)", qw[X Y Z])
> ```
> is the string `"\0X\0\0YZ"`.

Because of the double quotes, the `@1`, `@2`, and `@3` parts of the string are now interpreted as arrays to interpolate.¹  As Tony wrote, to make the example work with double quotes, you would have to escape all the `@` signs: `"\@1A((\@2A)\@3A)"`.

There's nothing really to test here because we don't have automatic verification of example code. The best we could do would be to copy this code into a `t` file, but that would not prevent future POD edits from accidentally breaking the example again. Also, the functionality in question is already covered by a test in `t/op/pack.t` (line 1163-1164), which contains almost identical code (using single quotes, of course): https://github.com/Perl/perl5/blob/v5.40.0/t/op/pack.t#L1163-L1164

---

¹ They're empty, so nothing shows up in the result, i.e. the template is effectively just `'A((A)A)'`. And as digit variables they're exempt from `strict 'vars'`.

khwilliamson commented:
My assertion is that it is better to standardize on a 1 space verbatim indent.  That is necessary, and sufficient.  Larger indents increase the chance of wrapping text that is too wide for the window it is being displayed in, making it harder to read.  Perl aims to try to prevent this, with only necessary and rare exceptions, in windows that are as small as the traditional 80 columns, with no attempt to accommodate narrower windows.

Even if the current text all fits, it places an unnecessary future maintenance burden on later developers to fit additions into a space that doesn't actually have to be that narrow.  I myself have wrestled with what to do upon encountering that situation: to lower the indent on all verbatim blocks of a file to preserve consistency, or just the ones that have been pushed over the edge.

I also don't see the point of an increased indent.  AFAIK it matters only when looking at the file with an editor.  A normal user would be viewing it as a man page or perldoc output or through a web browser.  Those all override the file's indentation to nicely set off the verbatim text from the rest.

mauke commented:
Wait, are you saying the number of spaces in a verbatim paragraph is semantically significant (i.e. affects the rendering)? If so, I'd consider that a bug in perldoc. I don't see it mentioned in perlpod or perlpodspec either. (In fact, perlpod implies the opposite: *"[...] so I could slurp in a working program, shift it over four spaces, and have it print out, er, verbatim.  And presumably in a monospace font."*)

That said, if you want to normalize verbatim paragraphs in perlfunc to use 1-space indentation, feel free. By my count, the formatting of verbatim paragraphs in perlfunc is as follows:

| initial indent | count |
|---------------:|------:|
|  4 spaces      | 444   |
|  2 spaces      |  24   |
|  1 space       |  22   |
|  3 spaces      |  18   |
|  8 spaces      |  13   |
|  1 tab         |   6   |
|  7 spaces      |   2   |
|  5 spaces      |   1   |
| 12 spaces      |   1   |
| 36 spaces      |   1   |

Currently 4-space indentation is the clear winner and I don't really want to touch the rest of perlfunc in this PR.

tonycoz commented on pod/perlfunc.pod:
```diff
@@ -5932,7 +5932,7 @@ L<C<unpack>|/unpack TEMPLATE,EXPR>, also via the C</>
 C<@> starts over at 0.  Therefore, the result of
 
-    pack("@1A((@2A)@3A)", qw[X Y Z])
+    pack('@1A((@2A)@3A)', qw[X Y Z])
```
We've already had someone who I'd expect to know better mess up the quotes here, maybe it could use a comment warning about that.

mauke replied:
Good point. Added.

tonycoz commented:
I'm not fussy about the consistent verbatim paragraph indentation, but if it's being done I do think it should be a separate commit to the substantive fix to the example.

mauke committed:
perlfunc/pack: indent verbatim paragraphs consistently

Previously, there was a mix of 1-space, 4-space, 3-space, 7-space, and
tab-based indentation. Most of perlfunc uses 4 spaces, so go with that.

Also, align the second column in the "This code:" ... "gives this result:"
table. To fit within the line length limit (and make podcheck.t happy),
use slightly shorter sample strings.

mauke committed:
perlfunc/pack: fix pack @ example

- Fix the `@1A((@2A)@3A)` example to use single quotes, not double
  quotes (which would interpolate array variables).
- Replace `pack L=>0x12345678` by the more obvious `pack("L",
  0x12345678)` (we're not trying to obfuscate our example code (right?)).

Fixes #22425.

mauke commented:
OK, I've split the indentation/alignment changes out in a separate commit.

khwilliamson commented:
The number of spaces is not semantically significant.  What I'm saying is that if the number is important enough to standardize it, the best number to standardize it to is 1, for the reasons I cited.

mauke commented:
Ah, I see. Sorry for the misunderstanding.

My point is that I have my editor configured for 4-space indentation, so that is the natural shift width to use for verbatim blocks as well. This is especially important for code examples using nested blocks (like loops and conditionals), which (with a 1-space base offset from POD) would have to be indented 5 and 9 (and generally 4n+1) spaces. Which constantly clashes with my default editor settings and requires special attention to avoid accidentally aligning to a 4n boundary again whenever I change or reformat code in POD.

Also, t/porting/podcheck.t assumes a base indent of 4 for verbatim paragraphs,¹ so if you use 4 spaces in the POD source, you can eyeball what podcheck.t is going to accept (at least in simple cases).

These are my arguments for using 4 spaces. It mainly comes down to editor ergonomics (laziness).

---

¹ Unless I'm misunderstanding the code, `sub end_Verbatim` checks whether `length(Text::Tabs::expand($lines[$i])) + $INDENT + $current_indent > $MAX_LINE_LENGTH` where `$lines[$i]` is the current line and `$current_indent` defaults to 0, but increases for every nested `=over`. `$INDENT` is hardcoded to 4.
